docs/source/includes/_query-parameters.md
# Query types

## Simple queries

> This query returns households by age, income group, and year for all census tracts in Massachusetts starting in 2010

```python
from strato_query import SDAPIQuery, APIQueryParams
from strato_query.filters import EqualToFilter, GreaterThanOrEqualToFilter

SDAPIQuery.query_api_df(
    APIQueryParams(
        data_fields=('year', 'geoid11', 'age_g', 'income_g', 'households'),
        table='incomeforecast_tract_annual_income_group_age',
        data_filters=[
            EqualToFilter('geoid2', 25),
            GreaterThanOrEqualToFilter('year', 2010),
        ],
        aggregations=(),
        groupby=()))
```

```r
library(stRatoquery)

df = submit_api_query(
  query = api_query_params(
    table = 'incomeforecast_tract_annual_income_group_age',
    data_fields = api_fields(fields_list = list('year', 'geoid11', 'age_g', 'income_g', 'households')),
    data_filters = list(
        eq_filter(filter_variable = 'geoid2', filter_value = 25),
        ge_filter(filter_variable = 'year', filter_value = 2010)
    ),
    aggregations=list(),
    groupby=list()
  ),
  apiToken = apiToken)
```

```shell
$ curl -X POST "https://api.stratodem.com/api" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{ \"token\": \"my-api-token\", \"query\": { \"query_type\": \"COUNT\", \"data_fields\": [ \"year\", \"geoid11\", \"age_g\", \"income_g\", \"households\" ], \"table\": \"incomeforecast_tract_annual_income_group_age\", \"groupby\": [ ], \"data_filters\": [ { \"filter_type\": \"eq\", \"filter_value\": 25, \"filter_variable\": \"geoid2\" }, { \"filter_type\": \"ge\", \"filter_value\": 2010, \"filter_variable\": \"year\" } ], \"aggregations\": [], \"order\": []}}"
```

```vb
Dim query As Dictionary

Set query = apiQueryParameters( _
    table:="incomeforecast_tract_annual_income_group_age", _
    dataFields:=Array("year", "geoid11", "age_g", "income_g", "households"), _
    dataFilters:=Array( _
        equalToFilter(filterVariable:="geoid2", filterValue:=25), _
        greaterThanOrEqualToFilter(filterVariable:="year", filterValue:=2010)), _
    groupby:=Array(), _
    aggregations:=Array())

' Write the results to "Example Sheet" (note that this sheet will need to exist first)
Call writeToSheet(results:=submitAPIQuery(query), sheetName:="Example Sheet")
```

Basic API queries evaluate like simple `SELECT` statements from the StratoDem Analytics database.

Queries roughly translate into SQL as `SELECT data_fields FROM table WHERE data_filters GROUP BY groupby;`

## Queries with aggregations

> This query returns the number of households age 75+ with at least $50,000 in household income by county and year for each county in California

```python
from strato_query import SDAPIQuery, APIQueryParams
from strato_query.aggregations import SumAggregation
from strato_query.filters import EqualToFilter, GreaterThanOrEqualToFilter

SDAPIQuery.query_api_df(
    APIQueryParams(
        # Passing in a dictionary to data_fields allows for renaming variables in the resulting DataFrame or JSON
        data_fields=['year', 'geoid5', {'households': 'households_age_75plus_hhi_50k_plus'}],
        table='incomeforecast_county_annual_income_group_age',
        data_filters=(
            EqualToFilter('geoid2', 6),
            # Cutoff for age 75+
            GreaterThanOrEqualToFilter('age_g', 16),
            # Cutoff for $50,000+ HHI
            GreaterThanOrEqualToFilter('income_g', 10),
        ),
        groupby=('year', 'geoid5'),
        aggregations=[SumAggregation('households')],
        order=('year', 'geoid5')
    ))
```

```r
library(stRatoquery)

df = submit_api_query(
  query = api_query_params(
    table = 'incomeforecast_county_annual_income_group_age',
    # Passing in a mapping to data_fields allows for renaming variables in the resulting data.frame or JSON
    data_fields = api_fields(fields_list = list('year', 'geoid5', list(households = 'households_age_75plus_50k_plus'))),
    data_filters = list(
        eq_filter(filter_variable = 'geoid2', filter_value = 6),
        # Cutoff for age 75+
        ge_filter(filter_variable = 'age_g', filter_value = 16),
        # Cutoff for $50,000+ HHI
        ge_filter(filter_variable = 'income_g', filter_value = 10)
    ),
    aggregations = list(sum_aggregation(variable_name = 'households')),
    groupby = list('year', 'geoid5'),
    order = list('year', 'geoid5')
  ),
  apiToken = apiToken)
```

```shell
$ curl -X POST "https://api.stratodem.com/api" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{ \"token\": \"my-api-token\", \"query\": {\"query_type\": \"COUNT\", \"data_fields\": [\"year\", \"geoid5\", {\"households\": \"households_age_75plus_hhi_50k_plus\"}], \"table\": \"incomeforecast_county_annual_income_group_age\", \"groupby\": [\"year\", \"geoid5\"], \"data_filters\": [{\"filter_type\": \"eq\", \"filter_value\": 6, \"filter_variable\": \"geoid2\"}, {\"filter_type\": \"ge\", \"filter_value\": 16, \"filter_variable\": \"age_g\"}, {\"filter_type\": \"ge\", \"filter_value\": 10, \"filter_variable\": \"income_g\"}], \"aggregations\": [{\"aggregation_func\": \"sum\", \"variable_name\": \"households\"}], \"order\": [\"year\", \"geoid5\"]}}"
```

```vb
Dim query As Dictionary

Set query = apiQueryParameters( _
    table:="incomeforecast_county_annual_income_group_age", _
    dataFields:=Array("year", "geoid5", renameVariable(original:="households", renamed:="households_age_75plus_hhi_50k_plus")), _
    dataFilters:=Array( _
        equalToFilter(filterVariable:="geoid2", filterValue:=6), _
        greaterThanOrEqualToFilter(filterVariable:="age_g", filterValue:=16), _
        greaterThanOrEqualToFilter(filterVariable:="income_g", filterValue:=10)), _
    groupby:=Array("year", "geoid5"), _
    order:=Array("year", "geoid5"), _
    aggregations:=Array(sumAggregation(variableName:="households")))

' Write the results to "Example Sheet" (note that this sheet will need to exist first)
Call writeToSheet(results:=submitAPIQuery(query), sheetName:="Example Sheet")
```

In many cases, we want to aggregate data returned by the API. For example, if we have the SQL query:
`SELECT year, geoid5, sum(households) FROM incomeforecast_county_annual_income_group_age WHERE income_g >= 5 AND age_g >= 17 GROUP BY year, geoid5 ORDER BY year, geoid5;`, 
we need to group by the `year` and `geoid5` (county) columns, and then get the sum of `households` for each combination.

The `groupby` and `aggregations` arguments tell the API which columns to group on and which aggregations to apply.

## Median queries

> This query returns the median household income for all households by census tract in Suffolk County, MA in 2019

```python
from strato_query import SDAPIQuery, APIMedianQueryParams
from strato_query.filters import EqualToFilter

SDAPIQuery.query_api_df(
    APIMedianQueryParams(
        median_variable_name='income_g',
        table='incomeforecast_tract_annual_income_group',
        data_fields=('year', 'geoid11', 'median_value'),
        data_filters=[
            EqualToFilter('year', 2019),
            EqualToFilter('geoid5', 25025),
        ],
        aggregations=(),
        groupby=('year', 'geoid11')       
    ))
```

```r
library(stRatoquery)

df = submit_api_query(
  query = median_query_params(
    table = 'incomeforecast_tract_annual_income_group',
    data_fields = api_fields(fields_list = list('year', 'geoid11', 'median_value')),
    data_filters = list(
        eq_filter(filter_variable = 'year', filter_value = 2019),
        eq_filter(filter_variable = 'geoid5', filter_value = 25025)
    ),
    groupby=c('year', 'geoid11'),
    median_variable_name='income_g',
    aggregations=list()
  ),
  apiToken = apiToken)
```

```shell
$ curl -X POST "https://api.stratodem.com/api" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{ \"token\": \"my-api-token\", \"query\": {\"query_type\": \"MEDIAN\", \"data_fields\": [\"year\", \"geoid11\", \"median_value\"], \"table\": \"incomeforecast_tract_annual_income_group\", \"groupby\": [\"year\", \"geoid11\"], \"data_filters\": [{\"filter_type\": \"eq\", \"filter_value\": 2019, \"filter_variable\": \"year\"}, {\"filter_type\": \"eq\", \"filter_value\": 25025, \"filter_variable\": \"geoid5\"}], \"aggregations\": [], \"median_variable_name\": \"income_g\"}}"
```

```vb
Dim query As Dictionary

Set query = medianQueryParameters( _
    table:="incomeforecast_tract_annual_income_group", _
    dataFields:=Array("year", "geoid11", "median_value"), _
    dataFilters:=Array(equalToFilter(filterVariable:="year", filterValue:=2019), equalToFilter(filterVariable:="geoid5", filterValue:=25025)), _
    aggregations:=Array(), _
    groupby:=Array("year", "geoid11"), _
    medianVariableName:="income_g")

' Write the results to "Example Sheet" (note that this sheet will need to exist first)
Call writeToSheet(results:=submitAPIQuery(query), sheetName:="Example Sheet")
```

To query median household income, net worth, home values, or other metrics, a special median query is needed.

These queries require a `median_variable_name` argument, which tells the API service which variable to 
use to compute the median value.

## Mean/average queries

> This query returns the median household income for all households by census tract in Suffolk County, MA in 2019

```python
from strato_query import SDAPIQuery, APIMeanQueryParams
from strato_query.filters import EqualToFilter

SDAPIQuery.query_api_df(
    APIMeanQueryParams(
        mean_variable_name='income_g',
        table='incomeforecast_tract_annual_income_group',
        data_fields=('year', 'geoid11', 'mean_value'),
        data_filters=[
            EqualToFilter('year', 2019),
            EqualToFilter('geoid5', 25025),
        ],
        aggregations=(),
        groupby=('year', 'geoid11')       
    ))
```

```r
library(stRatoquery)

df = submit_api_query(
  query = mean_query_params(
    table = 'incomeforecast_tract_annual_income_group',
    data_fields = api_fields(fields_list = list('year', 'geoid11', 'mean_value')),
    data_filters = list(
        eq_filter(filter_variable = 'year', filter_value = 2019),
        eq_filter(filter_variable = 'geoid5', filter_value = 25025)
    ),
    groupby = c('year', 'geoid11'),
    mean_variable_name = 'income_g',
    aggregations = list()
  ),
  apiToken = apiToken)
```

```shell
$ curl -X POST "https://api.stratodem.com/api" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{ \"token\": \"my-api-token\", \"query\": {\"query_type\": \"MEAN\", \"data_fields\": [\"year\", \"geoid11\", \"mean_value\"], \"table\": \"incomeforecast_tract_annual_income_group\", \"groupby\": [\"year\", \"geoid11\"], \"data_filters\": [{\"filter_type\": \"eq\", \"filter_value\": 2019, \"filter_variable\": \"year\"}, {\"filter_type\": \"eq\", \"filter_value\": 25025, \"filter_variable\": \"geoid5\"}], \"aggregations\": [], \"mean_variable_name\": \"income_g\"}}"
```

```vb
Dim query As Dictionary

Set query = meanQueryParameters( _
    table:="incomeforecast_tract_annual_income_group", _
    dataFields:=Array("year", "geoid11", "mean_value"), _
    dataFilters:=Array(equalToFilter(filterVariable:="year", filterValue:=2019), equalToFilter(filterVariable:="geoid5", filterValue:=25025)), _
    aggregations:=Array(), _
    groupby:=Array("year", "geoid11"), _
    meanVariableName:="income_g")

' Write the results to "Example Sheet" (note that this sheet will need to exist first)
Call writeToSheet(results:=submitAPIQuery(query), sheetName:="Example Sheet")
```

To query mean household income, net worth, home values, or other metrics, a special mean query is needed.

These queries require a `mean_variable_name` argument, which tells the API service which variable to 
use to compute the mean value.

## Combining multiple queries

> This query joins two results on common columns to compute population density

```python
from strato_query import SDAPIQuery, APIQueryParams
from strato_query.filters import EqualToFilter, LessThanFilter

df = SDAPIQuery.query_api_df(
    query_params=APIQueryParams(
        query_type='COUNT',
        table='populationforecast_metro_annual_population',
        data_fields=('year', 'cbsa', {'population': 'population'}),
        data_filters=(
            LessThanFilter(var='year', val=2015),
            EqualToFilter(var='cbsa', val=14454),
        ),
        aggregations=(dict(aggregation_func='sum', variable_name='population'),),
        groupby=('cbsa', 'year'),
        order=('year',),
        join=APIQueryParams(
            query_type='AREA',
            table='geocookbook_metro_na_shapes_full',
            data_fields=('cbsa', 'area', 'name'),
            data_filters=[EqualToFilter(var='cbsa', val=14454)],
            groupby=('cbsa', 'name'),
            aggregations=(),
            on=dict(left=('cbsa',), right=('cbsa',)),
        )
    )
)

df['POP_PER_SQ_MI'] = df['POPULATION'].div(df['AREA'])
df_final = df[['YEAR', 'NAME', 'POP_PER_SQ_MI']]
```

```r
library(stRatoquery)

api_query_params(
  table = 'populationforecast_metro_annual_population',
  data_fields = api_fields(fields_list = list('year', 'cbsa', list('population' = 'population'))),
  data_filters = list(lt_filter(filter_variable = "year", filter_value = 2015), eq_filter(filter_variable = "cbsa", filter_value = 14454)),
  aggregations = list(sum_aggregation(variable_name = "population")),
  groupby = c("cbsa", "year"),
  order = c("year"),
  join = api_query_params(
      table = 'geocookbook_metro_na_shapes_full',
      data_fields = api_fields(fields_list = list('cbsa', 'area', 'name')),
      data_filters = list(eq_filter(filter_variable = "cbsa", filter_value = 14454)),
      aggregations = list(),
      groupby = c("cbsa", "name"),
      on = list(left = c('cbsa'), right = c('cbsa')),
      query_type = "AREA"),
  query_type = "COUNT")
```

```shell
$ curl -X POST "https://api.stratodem.com/api" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{ \"token\": \"my-api-token\", \"query\": {\"query_type\": \"COUNT\", \"data_fields\": [\"year\", \"cbsa\", {\"population\": \"population\"}], \"table\": \"populationforecast_metro_annual_population\", \"groupby\": [\"cbsa\", \"year\"], \"data_filters\": [{\"filter_type\": \"lt\", \"filter_value\": 2015, \"filter_variable\": \"year\"}, {\"filter_type\": \"eq\", \"filter_value\": 14454, \"filter_variable\": \"cbsa\"}], \"aggregations\": [{\"aggregation_func\": \"sum\", \"variable_name\": \"population\"}], \"join\": {\"query_type\": \"AREA\", \"data_fields\": [\"cbsa\", \"area\", \"name\"], \"table\": \"geocookbook_metro_na_shapes_full\", \"groupby\": [\"cbsa\", \"name\"], \"data_filters\": [{\"filter_type\": \"eq\", \"filter_value\": 14454, \"filter_variable\": \"cbsa\"}], \"aggregations\": [], \"on\": {\"left\": [\"cbsa\"], \"right\": [\"cbsa\"]}}, \"order\": [\"year\"]}}"
```

```vb
apiQueryParameters( _
    table:="populationforecast_metro_annual_population", _
    dataFields:=Array("year", "cbsa", renameVariable(original:="population", renamed:="population")), _
    dataFilters:=Array(lessThanFilter(filterVariable:="year", filterValue:=2015), equalToFilter(filterVariable:="cbsa", filterValue:=14454)), _
    aggregations:=Array(sumAggregation(variableName:="population")), _
    groupby:=Array("cbsa", "year"), order:=Array("year"), _
    join:=apiQueryParameters( _
        table:="geocookbook_metro_na_shapes_full", _
        dataFields:=Array("cbsa", "area", "name"), _
        dataFilters:=Array(equalToFilter(filterVariable:="cbsa", filterValue:=14454)), _
        aggregations:=Array(), _
        groupby:=Array("cbsa", "name"), _
        joinOn:=joinOnStructure(left:=Array("cbsa"), right:=Array("cbsa")), _
        queryType:="AREA"), _
    queryType:="COUNT")
```

In many cases, we need to derive data from two or more variables. For example, when calculating 
population density, we need to take the population and divide by the area. In this case, those two 
metrics come from different sources in the StratoDem Analytics database, and we need to `JOIN` two queries 
using the `join` argument. 

A `join` takes another `APIQueryParams` object to join (effectively joining the results of two SELECT statements),
with that new `APIQueryParams` object requiring an `on` argument, to tell the API service which column(s) to join on.

For example, `on={"left": [cbsa], "right": [cbsa]}` tells the API to join the two subqueries on the `cbsa` column.

## GeoJSON queries

> This query returns a GeoJSON FeatureCollection with median household income stored in properties for all census tracts within a 15-minute drive of the lat-lng

```python
from strato_query import SDAPIQuery, APIGeoJSONQueryParams, APIMedianQueryParams
from strato_query.filters import EqualToFilter, DrivetimeFilter

geojson = SDAPIQuery.query_api_json(
    APIGeoJSONQueryParams(
        properties=[
            'geoid11',
            'name',
            'median_hhi'
        ],
        data_fields=['geoid11', 'geometry', 'name'],
        table='geocookbook_tract_na_shapes_full',
        groupby=(),
        data_filters=(),
        aggregations=(),
        join=APIMedianQueryParams(
            on={'left': ['geoid11'], 'right': ['geoid']},
            table='incomeforecast_tract_annual_income_group',
            median_variable_name='income_g',
            data_fields=[{'geoid11': 'geoid'}, {'median_value': 'median_hhi'}],
            data_filters=[
                EqualToFilter('year', 2019),
                DrivetimeFilter(latitude=42.1, longitude=-120.1, minutes=15, detailed_type='drivetime_unweighted'),
            ],
            groupby=['geoid11'],
            aggregations=(),
        )
    ))
```

```r
```

```shell
$ curl -X POST "https://api.stratodem.com/api" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{ \"token\": \"my-api-token\", \"query\": {\"query_type\": \"GEOJSON\", \"data_fields\": [\"geoid11\", \"geometry\", \"name\"], \"table\": \"geocookbook_tract_na_shapes_full\", \"groupby\": [], \"data_filters\": [], \"aggregations\": [], \"join\": {\"query_type\": \"MEDIAN\", \"data_fields\": [{\"geoid11\": \"geoid\"}, {\"median_value\": \"median_hhi\"}], \"table\": \"incomeforecast_tract_annual_income_group\", \"groupby\": [\"geoid11\"], \"data_filters\": [{\"filter_type\": \"eq\", \"filter_value\": 2019, \"filter_variable\": \"year\"}, {\"filter_type\": \"drivetime_unweighted\", \"filter_value\": {\"latitude\": 42.1, \"longitude\": -120.1, \"minutes\": 15, \"traffic\": \"disabled\", \"start_time\": null}, \"filter_variable\": \"\"}], \"aggregations\": [], \"on\": {\"left\": [\"geoid11\"], \"right\": [\"geoid\"]}, \"median_variable_name\": \"income_g\"}, \"properties\": [\"geoid11\", \"name\", \"median_hhi\"]}}"
```

```vb
```

To return a GeoJSON `FeatureCollection`, use the `APIGeoJSONQueryParams`. This query structure adds the 
`properties` argument, which specifies which fields will show up in each `Feature`'s `properties` field in the GeoJSON response.

The example to the right returns the GeoJSON `FeatureCollection` for the one census tract within a 15-minute drive of 42.1,-120.1 (lat, lng).

`{"type": "FeatureCollection", "features": [{"type": "Feature", "geometry": {"type": "Polygon", "coordinates": ["DUMMY DATA HERE"]}, "properties": {"geoid11": 41037960100, "name": "Census Tract: 41037960100", "median_hhi": 42182}}]}`

## Geocoder queries

> This query returns the census tract containing the latitude-longitude coordinates

```python
from strato_query import SDAPIQuery, APIGeocoderQueryParams

df_tract = SDAPIQuery.query_api_df(
    APIGeocoderQueryParams(
        data_fields=('geoid11', 'name'),
        # The geometry table for census tracts
        table='geocookbook_tract_na_shapes_full',
        latitude=42.3498224,
        longitude=-71.0521391,
        groupby=(),
        data_filters=(),
        aggregations=(),
    )
)

df_metro = SDAPIQuery.query_api_df(
    APIGeocoderQueryParams(
        data_fields=('cbsa', 'name'),
        # The geometry table for metros
        table='geocookbook_metro_na_shapes_full',
        latitude=42.3498224,
        longitude=-71.0521391,
        groupby=(),
        data_filters=(),
        aggregations=(),
    )
)
```

```r
# Not implemented yet
```

```shell
$ curl -X POST "https://api.stratodem.com/api" \
    -H "accept: application/json" \
    -H "Content-Type: application/json" \
    -d "{ \"token\": \"my-api-token\", \"query\": {\"query_type\": \"GEOCODER\", \"data_fields\": [\"cbsa\", \"name\"], \"table\": \"geocookbook_metro_na_shapes_full\", \"groupby\": [], \"data_filters\": [], \"aggregations\": [], \"latitude\": 42.3498224, \"longitude\": -71.0521391}}"
```

```vb
' Not implemented yet
```

To identify which geography contains a given location, the `APIGeocoderParams` take a 
geographic shapes table, a latitude, and a longitude, and construct a query to return the 
geographic ID.

The first example to the right returns the census tract ID (`GEOID11`) for the latitude-longitude pair.

|GEOID11|NAME|
|-------|----|
|25025061200|Census Tract: 25025061200|

The second example returns the metro ID for the latitude-longitude pair:

|CBSA|NAME|
|----|----|
|14454|Boston, MA|

## Custom polygon queries

> This query returns the population contained within the custom polygon by year

```python
from strato_query import SDAPIQuery, APIQueryParams
from strato_query.filters import IntersectsFilter, BetweenFilter
from strato_query.aggregations import SumAggregation

df = SDAPIQuery.query_api_df(
    APIQueryParams(
        table='populationforecast_tract_annual_population',
        data_filters=(
            IntersectsFilter(
                detailed_type='intersects_weighted',
                var='geoid11',
                val={"type": "Polygon",
                     "coordinates": [
                         [[-71.17801666259767, 42.43321295705304],
                          [-71.0145950317383, 42.43321295705304],
                          [-71.0145950317383, 42.3145122534915],

docsv2/source/includes/_query-parameters.md
# Query parameters


CHANGELOG.md
# Changelog for strato-query

All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [3.4.0] - 2019-11-06
### Adds
- Adds new `detailed_type` option to `IntersectsFilter`. It now supports:
    - `'intersects'` (simple intersection filtering)
    - `'intersects_weighted'` (ability to create weighted polygon results)

## Related issues
- [85](https://github.com/StratoDem/strato-query/issues/85)

## [3.3.0] - 2019-09-18
### Adds
- Adds new buffer types to the `DrivetimeFilter`. It now supports:
    - `'drivetime'`
    - `'drivetime_simple'`
    - `'drivetime_unweighted'`
    - `'drivetime_destination'`
    - `'drivetime_destination_simple'`
    - `'drivetime_destination_unweighted'`
- Adds a `WalktimeFilter` which supports the following types:
    - `'walktime'`
    - `'walktime_unweighted'`
    - `'walktime_destination'`
    - `'walktime_destination_unweighted'`
- Adds `detailed_type` as a new optional argument to `OverlapsDrivetimeFilter` which allows:
    - `'overlaps_drivetime'`  
    - `'overlaps_drivetime_destination'`
- Adds `detailed_type` as a new optional argument to `OverlapsWalktimeFilter` which allows:
    - `'overlaps_walktime'`  
    - `'overlaps_walktime_destination'`  

## Related issues
- [69](https://github.com/StratoDem/strato-query/issues/69)

## [3.2.0] - 2019-09-10
### Adds
- Adds new query param classes
    - `APIPureShapeQueryParams`
    - `APIPureShapeUnionQueryParams`
- Adds examples of new query params classes

## Related issues
- [69](https://github.com/StratoDem/strato-query/issues/69)

## [3.1.0] - 2019-09-10
### Adds
- Adds new filter types for overlaps. Each one takes the data needed to create the shape used to check against the specified geometry column
    - `OverlapsMileRadiusFilter`
    - `OverlapsDrivetimeFilter`
    - `OverlapsWalktimeFilter`

## Related issues
- [69](https://github.com/StratoDem/strato-query/issues/69)

## [3.0.2] - 2019-09-05
### Fixes
- Fixes broken links to examples in documentation. Now links to new docs site at code.stratodem.com

## [3.0.1] - 2019-08-26
### Changes
- Failed API queries no longer include API token value in the error message (replaced by `**********`)

## [3.0.0] - 2019-08-22
### Changes
- **Breaking** Renames filters to full words in Python package. E.g., `EqFilter` --> `EqualToFilter`
- Arguments no longer must be `tuple`s, but can also be `list`s :tada:
- `data_filters` argument for `APIQueryParams` may be an iterable of `BaseFilter` objects, instead of only `dict`s
- `BaseAggregation` and `SumAggregation` added to handle aggregations instead of relying on arbitrary `dict`s
- `aggregations` argument for `APIQueryParams` may be an iterable of `BaseAggregation` objects, instead of only `dict`s
- Renames `BaseAPIQuery` --> `SDAPIQuery`
- New authentication method to support environment variable at `STRATODEM_API_TOKEN` or `authenticate_to_api('my-api-token')`
- New documentation site with Python, R, Shell, and VBA examples :tada:

```python
from strato_query import SDAPIQuery, APIQueryParams, EqualToFilter, LessThanFilter
from strato_query.authentication import authenticate_to_api

authenticate_to_api('my-api-token')

df = SDAPIQuery.query_api_df(
    query_params=APIQueryParams(
        query_type='COUNT',
        table='populationforecast_metro_annual_population',
        data_fields=('year', 'cbsa', {'population': 'population'}),
        data_filters=(
            LessThanFilter(var='year', val=2015),
            EqualToFilter(var='cbsa', val=14454),
        ),
        aggregations=(dict(aggregation_func='sum', variable_name='population'),),
        groupby=('cbsa', 'year'),
        order=('year',),
        join=APIQueryParams(
            query_type='AREA',
            table='geocookbook_metro_na_shapes_full',
            data_fields=('cbsa', 'area', 'name'),
            data_filters=(EqualToFilter(var='cbsa', val=14454),),
            groupby=('cbsa', 'name'),
            aggregations=(),
            on=dict(left=('cbsa',), right=('cbsa',)),
        )
    )
)
```

## [2.11.1] - 2019-05-16
### Fixes
- Fixes an incorrect assertion statement in the `submit_query` method

### Related issues:
- [#53](https://github.com/StratoDem/strato-query/issues/53)

## [2.11.0] - 2019-05-07
### Adds
- Adds a new filter, `IntersectsFilter`, for querying all geoids whose geometries intersect with the given bounding geometry

### Related issues:
- [#50](https://github.com/StratoDem/strato-query/issues/50)

## [2.10.2] - 2019-04-22
### Changes
- Changes the pretty print methods in the `APIQueryParams` class so they can handle pretty printing for all query param classes
- Removes the pretty print methods from the `APICalculationQueryParams` class

### Related issues:
- [#48](https://github.com/StratoDem/strato-query/issues/48)

## [2.10.1] - 2019-04-22
### Changes
- Changes default timeout value to 60s

## [2.10.0] - 2019-04-22
### Adds
- Adds the `APICalculationQueryParams` class
- Adds the `APIFilterQueryParams` class

### Related issues:
- [#43](https://github.com/StratoDem/strato-query/issues/43)

## [2.9.0] - 2019-04-17
### Adds
- Adds automatic retries for the POST request if there is a connection issue
- Adds an optional, adjustable `timeout` param to exposed methods
- Adds docstrings to class methods

### Related issues:
- [#39](https://github.com/StratoDem/strato-query/issues/39)

## [2.8.0] - 2019-03-22
### Fixes
- Includes `APIGeoJSONQueryParams` in `__all__` for `base_API_query.py`

### Adds
- Adds `APIGeocoderQueryParams`, which, when called form
  `BaseAPIQuery.query_api_df`, returns a `DataFrame` with the geographic
  information associated with a latitude-longitude pair.
  
  Example: 
  ```python 
  from strato_query.base_API_query import BaseAPIQuery, APIGeocoderQueryParams

  BaseAPIQuery.query_api_df(
    query_params=APIGeocoderQueryParams(
        data_fields=('geoid11', 'geoid5',),

R/params.R
library(jsonlite)

#' This function structures an API query to submit to the StratoDem Analytics API
#' @param table Table name
#' @param query_type Type of query - defaults to 'COUNT'
#' @param data_fields List of fields to return
#' @param data_filters List of filters for the query
#' @param aggregations List of aggregations for the query
#' @param groupby List of variables for GROUP BY
#' @param order List of variables to ORDER
#' @keywords query
#' @export
#' @return query parameters
#' @examples
#' # Returns 80+ population for Massachusetts and California from 2010 onward, sorted by year and state ID
#' api_query_params(
#'   table = 'populationforecast_state_annual_population_age',
#'   data_fields = api_fields(fields_list = list('year', 'geoid2', list(population = 'population_80_plus'))),
#'   data_filters = list(
#'     ge_filter(filter_variable = 'age_g', filter_value = 17),
#'     in_filter(filter_variable = 'geoid2', filter_value = c(6, 25)),
#'     ge_filter(filter_variable = 'year', filter_value = 2010)),
#'   aggregations = list(sum_aggregation(variable_name = 'population')),
#'   groupby = c('year', 'geoid2'),
#'   order = c('year', 'geoid2'))
api_query_params = function(table, query_type = 'COUNT', data_fields, data_filters, aggregations = list(), groupby = list(),
                            order = list(), median_variable_name = NA, mean_variable_name = NA, join_type = 'JOIN', join = NA, on = NA) {
  if (is.na(on) && is.na(join)) {
    query = list(
      table = jsonlite::unbox(table),
      query_type = jsonlite::unbox(query_type),
      data_fields = data_fields,
      data_filters = data_filters,
      aggregations = aggregations,
      groupby = groupby,
      order = order,
      median_variable_name = jsonlite::unbox(median_variable_name),
      mean_variable_name = jsonlite::unbox(mean_variable_name),
      join = jsonlite::unbox(NA),
      on = jsonlite::unbox(NA)
    )
  } else if (is.na(on) && !is.na(join)) {
    query = list(
      table = jsonlite::unbox(table),
      query_type = jsonlite::unbox(query_type),
      data_fields = data_fields,
      data_filters = data_filters,
      aggregations = aggregations,
      groupby = groupby,
      order = order,
      median_variable_name = jsonlite::unbox(median_variable_name),
      mean_variable_name = jsonlite::unbox(mean_variable_name),
      join_type = join_type,
      join = join,
      on = jsonlite::unbox(NA)
    )
  } else if (!is.na(on) && is.na(join)) {
    query = list(
      table = jsonlite::unbox(table),
      query_type = jsonlite::unbox(query_type),
      data_fields = data_fields,
      data_filters = data_filters,
      aggregations = aggregations,
      groupby = groupby,
      order = order,
      median_variable_name = jsonlite::unbox(median_variable_name),
      mean_variable_name = jsonlite::unbox(mean_variable_name),
      join = jsonlite::unbox(NA),
      on = on
    )
  } else {
    query = list(
      table = jsonlite::unbox(table),
      query_type = jsonlite::unbox(query_type),
      data_fields = data_fields,
      data_filters = data_filters,
      aggregations = aggregations,
      groupby = groupby,
      order = order,
      median_variable_name = jsonlite::unbox(median_variable_name),
      mean_variable_name = jsonlite::unbox(mean_variable_name),
      join_type = join_type,
      join = join,
      on = on
    )
  }

  return(query)
}

#' This function structures an API query that gets the median value of the target "median_variable_name" to submit to the StratoDem Analytics API
#' @param table Table name
#' @param median_variable_name Name of median variable to use
#' @param data_fields List of fields to return
#' @param data_filters List of filters for the query
#' @param aggregations List of aggregations for the query
#' @param groupby List of variables for GROUP BY
#' @param order List of variables for ORDER
#' @keywords query
#' @export
#' @return query parameters
#' @examples
#' # Returns median household income by county in 2018
#' api_query_params(
#'   table = 'incomeforecast_county_annual_income_g',
#'   data_fields = api_fields(fields_list = list('geoid5', list(median_value = 'median_hhi'))),
#'   data_filters = list(eq_filter(filter_variable = 'year', filter_value = 2018)),
#'   aggregations = list(),
#'   median_variable_name = 'income_g',
#'   groupby = c('geoid5'),
#'   order = c('geoid5'))
median_query_params = function(table, data_fields, data_filters, median_variable_name, aggregations = list(),
                               groupby = list(), order = list(), join_type = 'JOIN', join = NA, on = NA) {
  return(api_query_params(
    table = table,
    query_type = 'MEDIAN',
    data_fields = data_fields,
    data_filters = data_filters,
    aggregations = aggregations,
    groupby = groupby,
    order = order,
    median_variable_name = median_variable_name,
    join_type = join_type,
    join = join,
    on = on
  ))
}

#' This function structures an API query that gets the average value of the target "mean_variable_name" to submit to the StratoDem Analytics API
#' @param table Table name
#' @param mean_variable_name Name of mean variable to use
#' @param data_fields List of fields to return
#' @param data_filters List of filters for the query
#' @param aggregations List of aggregations for the query
#' @param groupby List of variables for GROUP BY
#' @keywords query
#' @export
#' @return query parameters
#' @examples
#' # Returns average household income by state in 2018
#' api_query_params(
#'   table = 'incomeforecast_state_annual_mean_household_income',
#'   data_fields = api_fields(fields_list = list('geoid2', list(mean_value = 'mean_hhi'))),
#'   data_filters = list(eq_filter(filter_variable = 'year', filter_value = 2018)),
#'   aggregations = list(),
#'   mean_variable_name = 'mean_household_income',
#'   groupby = c('geoid2'),
#'   order = c('geoid2'))
mean_query_params = function(table, data_fields, data_filters, mean_variable_name, aggregations = list(), groupby = list(),
                             order = list(), join_type = 'JOIN', join = NA, on = NA) {
  return(api_query_params(
    table = table,
    query_type = 'MEAN',
    data_fields = data_fields,
    data_filters = data_filters,
    aggregations = aggregations,
    groupby = groupby,
    mean_variable_name = mean_variable_name,
    order = order,
    join_type = join_type,
    join = join,
    on = on
  ))
}
